mono/metadata/reflection.c

```diff
@@ -10715,19 +10715,23 @@ mono_reflection_register_with_runtime (MonoReflectionType *type)
  * LOCKING: Assumes the loader lock is held.
  */
 static MonoMethodSignature*
-parameters_to_signature (MonoImage *image, MonoArray *parameters) {
-	MonoError error;
+parameters_to_signature (MonoImage *image, MonoArray *parameters, MonoError *error) {
 	MonoMethodSignature *sig;
 	int count, i;
 
+	mono_error_init (error);
+
 	count = parameters? mono_array_length (parameters): 0;
 
 	sig = (MonoMethodSignature *)image_g_malloc0 (image, MONO_SIZEOF_METHOD_SIGNATURE + sizeof (MonoType*) * count);
 	sig->param_count = count;
 	sig->sentinelpos = -1; /* FIXME */
 	for (i = 0; i < count; ++i) {
-		sig->params [i] = mono_type_array_get_and_resolve (parameters, i, &error);
-		mono_error_raise_exception (&error); /* FIXME don't raise here */
+		sig->params [i] = mono_type_array_get_and_resolve (parameters, i, error);
+		if (!is_ok (error)) {
+			image_g_free (image, sig);
```

Comment:
`image_g_free` is something I added in 6052e0735345286f1f85ec2b7ee7b528324716f4  it's just a wrapper around `g_free` when `image == NULL`, otherwise it's a no-op.  It doesn't actually dispose `image` itself.  (`image_g_malloc0` is just malloc when `image` is NULL, so I'm only cleaning up in the case when it's not allocating into an image)